config-service-impl/src/main/java/org/hypertrace/config/service/store/DocumentConfigStore.java

```diff
@@ -46,34 +49,54 @@ public class DocumentConfigStore implements ConfigStore {
 
   private final Datastore datastore;
   private final Collection collection;
+  private final FilterBuilder filterBuilder;
 
   public DocumentConfigStore(Clock clock, Datastore datastore) {
     this.clock = clock;
     this.datastore = datastore;
     this.collection = this.datastore.getCollection(CONFIGURATIONS_COLLECTION);
+    this.filterBuilder = new FilterBuilder();
   }
 
   @Override
   public UpsertedConfig writeConfig(
       ConfigResourceContext configResourceContext,
       String lastUpdatedUserId,
-      Value latestConfig,
+      UpsertConfigRequest request,
       String lastUpdatedUserEmail)
       throws IOException {
     Optional<ConfigDocument> previousConfigDoc = getLatestVersionConfigDoc(configResourceContext);
     Optional<ContextSpecificConfig> optionalPreviousConfig =
         previousConfigDoc.flatMap(this::convertToContextSpecificConfig);
 
+    // reject create config with condition
+    if (optionalPreviousConfig.isEmpty() && request.hasUpsertCondition()) {
+      throw Status.INVALID_ARGUMENT
+          .withDescription("No upsert condition required for creating config")
+          .asRuntimeException();
+    }
+
     Key latestDocKey = new ConfigDocumentKey(configResourceContext);
     ConfigDocument latestConfigDocument =
         buildConfigDocument(
             configResourceContext,
-            latestConfig,
+            request.getConfig(),
             lastUpdatedUserId,
             previousConfigDoc,
             lastUpdatedUserEmail);
 
-    collection.upsert(latestDocKey, latestConfigDocument);
+    if (request.hasUpsertCondition()) {
+      Filter filter = this.filterBuilder.buildDocStoreFilter(request.getUpsertCondition());
+      UpdateResult updateResult = collection.update(latestDocKey, latestConfigDocument, filter);
+      if (updateResult.getUpdatedCount() <= 0) {
+        throw Status.FAILED_PRECONDITION
+            .withDescription("Update failed because upsert condition did not match given record")
+            .asRuntimeException();
+      }
+    } else {
+      collection.createOrReplace(latestDocKey, latestConfigDocument);
```

Comment:
Out of curiosity, what's the difference between upsert and createOrReplace?

Reply:
1. The upsert api mentions to use createOrReplace
2. Both uses same mongo api `updateOne` to write it. Only difference i see is with preparing the document, it had generic literal wrapping and some projection.
3. checked with suresh also if it's safe to replace method, they are using it already